### PR TITLE
Distributed workers sample

### PR DESCRIPTION
**Pull request by chbatey**
This includes the feedback from this PR:https://github.com/akka/akka-distributed-workers-scala.g8/pull/19/files

It now uses the receptionist

### CODE REVIEW

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+  val ResultsTopic = "results"
+
+  case class Ack(workId: String) extends CborSerializable
+
+  sealed trait WorkerStatus
+  case object Idle extends WorkerStatus
+  case class Busy(workId: String, deadline: Deadline) extends WorkerStatus
+  case class WorkerState(ref: ActorRef[Worker.Command], status: WorkerStatus)
```
Comment (johanandren) [inline akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala]: I think these should be placed above the behavior instead of lurking down here.

File: akka-sample-distributed-workers-scala/src/main/scala/worker/WorkResultConsumer.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.cluster.pubsub.DistributedPubSub
+import akka.cluster.pubsub.DistributedPubSubMediator
+import akka.actor.typed.scaladsl.adapter._
+
+object WorkResultConsumer {
+  def apply(): Behavior[Any] =
+    Behaviors.setup[Any] { ctx =>
+      val mediator = DistributedPubSub(ctx.system).mediator
```
Comment (johanandren) [inline akka-sample-distributed-workers-scala/src/main/scala/worker/WorkResultConsumer.scala]: Leave a fixme to use the typed pubsub when ready akka/akka#26338

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    }
+
+}
+
+/**
+ * The worker is actually more of a middle manager, delegating the actual work
+ * to the WorkExecutor, supervising it and keeping itself available to interact with the work master.
+ */
+object Worker {
```
Comment (johanandren) [inline akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala]: Maybe we should just rename it to `MiddleManager` then? (I'm at least half serious ;) )

A bit weird with the companion below the class IMO

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+  sealed trait Command extends CborSerializable
+  case object WorkIsReady extends Command
+  case class Ack(id: String) extends Command
+  case class SubmitWork(work: Work) extends Command
+  case class WorkComplete(result: String) extends Command
```
Comment (johanandren) [inline akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala]: 3 of these sounds more like events than commands to me, (4 including WorkTimeout)

File: akka-sample-distributed-workers-scala/worker.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# The Worker Nodes
+
+`Worker` actors and the `Master` actor interact as follows:
+
+1. `Worker` actors register with the `Master` so the master knows they are available and ready to take on work.
```
Comment (johanandren) [inline akka-sample-distributed-workers-scala/worker.md]: They register with the receptionist, the `Master` subscribes to the receptionist.

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+import worker.WorkState.WorkStarted
+import worker.WorkState.WorkerFailed
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.duration.{ Deadline, FiniteDuration, _ }
+
+/**
+ * The master actor keep tracks of all available workers, and all scheduled and ongoing work items
+ */
+object Master {
```
Comment (johanandren) [inline akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala]: Should we call it something else than master (not sure it is worth it but it does makes me think of master-slave) and it really is more of a `WorkManager` or something anyways or what is it the master of?

File: akka-sample-distributed-workers-scala/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+dumps
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/.gitignore]: Ops that should go in my global one. It's graal crashing

File: akka-sample-distributed-workers-scala/README.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+![Overview](images/cluster-nodes.png)
+
+Let's look at the details of each part of the application, starting with the front-end.
+
+* [The Front-end Nodes](front-end.md)
+* [The Back-end Nodes](back-end.md)
+* [The Master Actor in Detail](master-in-detail.md)
+* [The Worker Nodes](worker.md)
+* [Experimenting with the example](experimenting.md)
+* [Next Steps](next-steps.md)
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/README.md]: No longer published

File: akka-sample-distributed-workers-scala/back-end.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+The singleton accepts the `Behavior` of the actual singleton actor, as well as configuration 
+which allows us to decide that the singleton actors should only run on the nodes with the role `back-end`.
+
+Calls to `init` on nodes without the `back-end` role will result in a proxy to communicate with the singleton
+being created.
+
+The state of the master is recovered on the standby node in the case of the node being lost through event sourcing.
+
+An alternative to event sourcing and the singleton master would be to keep track of all jobs in a central database, but that is more complicated and not as scalable.
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/back-end.md]: Leftover from the previous version. I think i'll just remove that :)

File: akka-sample-distributed-workers-scala/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/README.md]: I hadn't ever noticed docs gen, i'll see where best to put these docs

File: akka-sample-distributed-workers-scala/src/main/resources/logback.xml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<configuration>
+
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <encoder>
+            <pattern>[%date{ISO8601}] [%level] [%logger] [%thread] [%X{akkaSource}] - %msg%n</pattern>
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/src/main/resources/logback.xml]: :+1:

File: akka-sample-distributed-workers-scala/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-sample-distributed-workers-scala/README.md]: I don't quite remember how it works, but I think each project that is published to techhub has a sister project in there that generates a paradox site out of the README.md, not sure it the other samples has more than one page like this so curious how that works out.

File: akka-sample-distributed-workers-scala/src/main/scala/worker/FrontEnd.scala [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+      ctx: ActorContext[Command]): Behavior[Command] = {
+    val nextTick = ThreadLocalRandom.current.nextInt(3, 10).seconds
+    timers.startSingleTimer("tick", Tick, nextTick)
+    Behaviors.receiveMessage {
+      case Tick =>
+        busy(workCounter + 1, Work(nextWorkId(), workCounter), masterProxy, timers, ctx)
+      case _ =>
+        Behaviors.unhandled
+    }
+  }
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/src/main/scala/worker/FrontEnd.scala]: Yep that'll be much nicer

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala (hunk 4 above)
Comment (chbatey) [reply akka-sample-distributed-workers-scala/src/main/scala/worker/Worker.scala]: Perhaps message then, don't want to confuse things with event sourced behaviors

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala (hunk 6 above)
Comment (chbatey) [reply akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala]: Yeah best to remove that kind of terminology

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+  val WorkerServiceKey: ServiceKey[Worker.Command] = ServiceKey[Worker.Command]("workerService")
+
+  sealed trait Command extends CborSerializable
+  // Messages from Workers
+  case class UpdatedWorkers(workers: Receptionist.Listing) extends Command
+  case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[Worker.Command]) extends Command
+  case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[Worker.Command])
+      extends Command
+  case class WorkFailed(worker: ActorRef[Worker.Command], workId: String) extends Command
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala]: I'm not sure about this. These two actors are only designed to work together. There won't be other actors that can register to receive work. In that case is it wrong to have the two actor's protocols coupled this way? What's the advantage of having separate types in the Master that are always converted to Worker messages?

[the remark this answers is not in the document]
Comment (johanandren) [reply akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala]: Even for component internals I think we shouldn't do circular dependencies with types, hard to follow and brittle. I agree adapting types feel like overhead, I think the types should be defined in one place so that we either have the dependency from worker to master or the other way around.

Having them with the master is probably most reasonably since it is like a protocol that the worker should support/implement to be able to work with the master?

File: akka-sample-distributed-workers-scala/src/main/resources/application.conf [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  actor {
+    # Must be set like this to use Akka Cluster
+    provider = cluster
+
+    serialization-bindings {
+      "worker.CborSerializable" = jackson-cbor
+    }
+  }
+
+  loglevel = DEBUG
```
Comment (patriknw) [inline akka-sample-distributed-workers-scala/src/main/resources/application.conf]: INFO?

File: akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala]: Okay i'll move them to the master

File: akka-sample-distributed-workers-scala/src/main/resources/application.conf (hunk 13 above)
Comment (chbatey) [reply akka-sample-distributed-workers-scala/src/main/resources/application.conf]: the logback is INFO

File: akka-sample-distributed-workers-scala/src/main/scala/worker/WorkManager.scala [hunk 14]
```diff
@@ -25,24 +25,30 @@ import scala.concurrent.duration.{ Deadline, FiniteDuration, _ }
...
+  final case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[WorkerCommand])
       extends Command
-  final case class WorkFailed(worker: ActorRef[Worker.Message], workId: String) extends Command
+  final case class WorkFailed(worker: ActorRef[WorkerCommand], workId: String) extends Command
+
+  // Responses to requests from workers
+  sealed trait WorkerCommand
+  final case class DoWork(work: Work) extends WorkerCommand
+  final case class WorkAck(id: String) extends WorkerCommand
+  case object WorkAvailable extends WorkerCommand
```
Comment (johanandren) [inline akka-sample-distributed-workers-scala/src/main/scala/worker/WorkManager.scala]: 👍

**Comment by patriknw** [comment]
there are more "master" if you want to completely rename that

**Comment by chbatey** [comment]
Another find and replace of Master->WorkManger, can be merged once done

**Comment by patriknw** [comment]
@chbatey could you also add link to sample in https://doc.akka.io/docs/akka/current/project/examples.html
